src/iqm_client/iqm_client.py

```diff
@@ -334,3 +430,80 @@ def wait_for_results(self, run_id: UUID, timeout_secs: float = DEFAULT_TIMEOUT_S
+
+        url = f'{self._credentials.auth_server_url}/realms/{AUTH_REALM}/protocol/openid-connect/logout'
+        data = AuthRequest(client_id=AUTH_CLIENT_ID, refresh_token=self._credentials.refresh_token)
+        result = requests.post(url, data=data.dict(exclude_none=True))
+        if result.status_code not in [200, 204]:
+            raise ClientAuthenticationError(f'Logout failed, {result.text}')
+        self._credentials.access_token = None
+        self._credentials.refresh_token = None
+        return True
+
+    def _get_bearer_token(self, retries: int = 1) -> Optional[str]:
+        """Make a bearer token for Authorization header. If access token is about to expire refresh it first.
+
+        Args:
+            retries: number of times to try updating the tokens
+
+        Returns:
+            Bearer token, i.e. string containing prefix 'Bearer ' and the access token, or None if access token
+            is not available.
+        """
+        if self._credentials is None or not self._credentials.access_token:
+            return None
+        if _time_left_seconds(self._credentials.access_token) > REFRESH_MARGIN_SECONDS:
+            # Access token is still valid, no need to refresh
+            return f'Bearer {self._credentials.access_token}'
+        if retries < 1:
+            return None
+        self._update_tokens()
+        return self._get_bearer_token(retries - 1)
+
+    def _update_tokens(self):
+        """Update access token and refresh token.
+
+        Uses refresh token to request new tokens from authentication server.  If the refresh token has expired or
+        is about to expire, starts a new session by requesting the tokens using username and password.
+
+        Updated tokens are stored in ``self._credentials``.
+
+        Raises:
+            ClientAuthenticationError: updating the tokens failed
+        """
+        if self._credentials is None:
+            return
+
+        refresh_token = self._credentials.refresh_token
+        if refresh_token and _time_left_seconds(refresh_token) > REFRESH_MARGIN_SECONDS:
+            # Update tokens using existing refresh_token
+            data = AuthRequest(
+                client_id=AUTH_CLIENT_ID,
+                grant_type=GrantType.REFRESH,
+                refresh_token=refresh_token
+            )
+        else:
+            # Update tokens using username and password
+            data = AuthRequest(
+                client_id=AUTH_CLIENT_ID,
+                grant_type=GrantType.PASSWORD,
+                username=self._credentials.username,
+                password=self._credentials.password
+            )
+
+        url = f'{self._credentials.auth_server_url}/realms/{AUTH_REALM}/protocol/openid-connect/token'
+        result = requests.post(url, data=data.dict(exclude_none=True))
+        if result.status_code != 200:
+            raise ClientAuthenticationError(f'Failed to update tokens, {result.text}')
+        tokens = json.loads(result.text)
```

Comment:
```suggestion
        tokens = result.json()
```
Maybe? Again, `requests` docs could be clearer on this.

Reply:
Also elsewhere in the module.

Reply:
Fixing this required a bit of tuning in tests, but now it's done.